# Exploratory Data Analysis (EDA)

## 1. Initial Procedures

### Importing Data
In this section, necessary Python libraries are imported, and the dataset containing all of the elements is read into the environment.

In [ ]:
# Import necessary libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Read the dataset
df = pd.read_csv('datasheet.csv')
# Delete all empty columns
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]

### Cleaning and Organizing Data

In this section, we reorganized and cleanted the dataset by deleting empty columns on the csv file, renaming columns for clarity, and converting some data to numeric to manipulate and work on the data easily.

The organized dataset contains 'DATE', 'TotalSleep_min', 'REM_min', 'Light_min', 'Deep_min', 'Caffeine_mg', and 'Productivity_min' information, where min indicates minutes.

In [ ]:
# Rename columns for clarity
df.columns = [
    'DATE', 'TotalSleep', 'TotalSleep_min', 'REM', 'REM_min',
    'Light', 'Light_min', 'Deep', 'Deep_min', 'Caffeine_mg',
    'Productivity', 'Productivity_min'
]

# Convert Caffeine to numeric
df['Caffeine_mg'] = df['Caffeine_mg'].str.replace('mg','').str.strip()
df['Caffeine_mg'] = pd.to_numeric(df['Caffeine_mg'], errors='coerce')

# Convert DATE to datetime format and sort
df['DATE'] = pd.to_datetime(df['DATE'] + ' 2025', format='%B %d %Y', errors='coerce')
df.sort_values('DATE', inplace=True)



## 2. Dataset Overview and Summary Statistics 

In this part, we compute the main statistical measures for our variables including count, mean, standard deviation and min/max values with the code beneath. 

In [ ]:

# Printing the summary statistics
stats = df[['Caffeine_mg', 'Productivity_min', 'TotalSleep_min',
            'REM_min', 'Light_min', 'Deep_min']].describe().T 
print(stats)


                  count    mean         std    min    25%    50%    75%    max
Caffeine_mg        25.0  114.76  113.518457    0.0   25.0  135.0  135.0  410.0
Productivity_min   25.0  196.60  100.485488   60.0  105.0  180.0  300.0  360.0
TotalSleep_min     25.0  438.60   52.605450  324.0  412.0  450.0  468.0  557.0
REM_min            25.0   95.28   23.129202   49.0   85.0   95.0  107.0  145.0
Light_min          25.0  288.68   42.891064  191.0  271.0  290.0  317.0  379.0
Deep_min           25.0   51.56   12.662543   33.0   39.0   52.0   59.0   75.0




**Findings from the Summary Statistics**

- The average daily caffeine intake is approximately 115 mg, with a standard deviation of 114 mg, indicating high variability. Some days had very little caffeine intake, while on other days it exceeded 400 mg.
- Total sleep time is fairly consistent, averaging 439 minutes (approximately 7.3 hours) with a standard deviation of 53 minutes.
- Sleep quality statistics:
  - REM sleep averages 95 minutes, with moderate variability.
  - Light sleep is the largest portion of sleep, averaging 289 minutes, with some days significantly higher or lower.
  - Deep sleep averages 52 minutes, being the most stable sleep stage with lower variability.

Overall, sleep times are relatively stable, whereas caffeine intake and productivity show wider fluctuations, suggesting that there could be a stronger correlation between caffeine consumption and productivity.


## 3. Correlation Analysis

This section investigates the relationships between daily caffeine consumption, total sleep duration, different sleep stages (indicating sleep quality), and productivity. A correlation matrix was computed for six variables: Caffeine_mg, TotalSleep_min, REM_min, Light_min, Deep_min, and Productivity_min. 

Main Insights from the Heatmap:
- `Productivity_min` is strongly positively correlated with `Caffeine_mg`
- `Productivity_min` has a moderate positive correlation with each of `TotalSleep_min`, `REM_min`, and `Light_min`.
- Surprisingly, a weak negative correlation exists between `Deep_min` and `Productivity_min`.

This analysis gives us a solid starting point to investigate and test hypotheses about how sleep and caffeine influence productivity.


In [ ]:
# CORRELATION MATRIX
all_corr = df[['TotalSleep_min', 'REM_min', 'Light_min', 'Deep_min', 'Caffeine_mg', 'Productivity_min']].corr()
print("Correlation Matrix for All Variables:\n", all_corr)
plt.figure(figsize=(8,6))
sns.heatmap(all_corr, annot=True, cmap='coolwarm', fmt=".2f")
plt.title('Correlation Matrix: Sleep, Caffeine & Productivity')
plt.show()


![Correlation Matrix](figures/correlationmatrix.jpg)



## 4. Visualization

###Daily Caffeine Intake Distribution
This histogram shows the distribution of daily caffeine consumption throughout the observation period, which is about a month.

The frequency of values indicates that caffeine intake was irregular. Most days fall within a lower to moderate consumption range, while a few days show remarkably high intake. The distribution is **skewed to the right**, meaning that lower caffeine consumption was more frequent, although there were occasional high values. The irregular peaks might reflect days with higher workload, such as exam periods, which could later help us explain the observed patterns of productivity.



In [ ]:
# Caffeine Consumption Distribution Histogram
plt.figure(figsize=(8,5))
sns.histplot(df['Caffeine_mg'], bins=10, color='red')
plt.xlim(left=0) #remove the empty gap between frequency axis' 0 and caffeine axis' 0
plt.title('Daily Caffeine Intake Distribution')
plt.xlabel('Caffeine (mg)')
plt.ylabel('Frequency')
plt.show()


![Caffeine Intake Distribution](figures/caffeine_distribution.jpg)




###Daily Sleep Distribution
This histogram shows the distribution of daily total sleep time (in minutes) recorded throughout the observation period.

From the histogram, it can be observed that most days fall within the 390-480 minute range, indicating that total sleep remained relatively consistent throughout the month. There are a few days with notably shorter or longer sleep durations, suggesting limited extremes. Overall, the distribution appears roughly symmetric, supporting the stabiltiy of sleep patterns over the period.



In [ ]:
# Total Sleep Distribution Histogram
plt.figure(figsize=(8,5))
sns.histplot(df['TotalSleep_min'], bins=10, color='yellowgreen')
plt.xlim(300, df['TotalSleep_min'].max())  #start the x axis from 300
plt.title('Daily Total Sleep Distribution')
plt.xlabel('Total Sleep (minutes)')
plt.ylabel('Frequency')
plt.show()


![Total Sleep Amount Distribution](figures/totalsleep_distribution.jpg)




###Caffeine vs Productivity
This scatter plot illustrates the relationship between daily caffeine consumption and productivity.

From the plot, such an inference can be made that higher caffeine intake generally corresponds to higher productivity, as most points trend upward while caffeine consumption increases. While most data points follow this trend, a few days show moderate or greater productivity even with lower caffeine intake, suggesting that other factors besides caffeine may also influence productivity. 



In [ ]:
# Caffeine vs Productivity scatter plot
plt.figure(figsize=(8,5))
sns.scatterplot(x='Caffeine_mg', y='Productivity_min', data=df)
plt.title('Caffeine vs Productivity')
plt.xlabel('Caffeine (mg)')
plt.ylabel('Productivity (minutes)')
plt.show()


![Caffeine vs Productivity](figures/caffeine_vs_productivity.jpg)




###Total Sleep vs Productivity
This scatter plot demonstrates the relationship between total sleep duration and productivity.

The plot illustrates that majority of the data points fall within 400-500 minutes range, revealing that the sleep pattern remained relatively stable in general. 
There appears to be a moderate positive trend: in general, days with longer sleep correspond to higher productivity. However, a notable amount of data points deviate from this pattern, suggesting that although the sleep duration may influence productivity, it is not the sole determining factor. Overall, the plot indicates that maintaining adequate sleep may support better productivity. 



In [ ]:
# Total sleep vs Productivity scatter plot 
plt.figure(figsize=(8,5))
sns.scatterplot(x='TotalSleep_min', y='Productivity_min', data=df)
plt.title('Total Sleep vs Productivity')
plt.xlabel('Total Sleep (minutes)')
plt.ylabel('Productivity (minutes)')
plt.show()


![Total Sleep vs Productivity](figures/totalsleep_vs_productivity.jpg)




###Productivity vs Caffeine Intake Over Time
This line graph visualizes the relationship between daily productivity and caffeine intake over time. The blue line represents productivity. while the red line illustrates caffeine intake over time. Both variables have been normalized to a 0 to 1 scale(0-1) to allow visual comparison despite their differing units. 

Both lines show some fluctuations throughout the month, indicating variability in daily habits. 
Peaks and troughs in productivity can be observed alongside fluctuations in caffeine consumption, which helps to identify patterns or potential correlations at a glance. For example, days with higher caffeine intake often appear to coincide with higher productivity levels, suggesting a positive relationship between the two variables.



In [ ]:
# Productivity vs Caffeine Intake Over Time
def normalize(col): #normalizing values to visually compare the 2 variables
    return (col - col.min()) / (col.max() - col.min())
plt.figure(figsize=(10,5))
plt.plot(df['DATE'], normalize(df['Productivity_min']), label='Productivity')
plt.plot(df['DATE'], normalize(df['Caffeine_mg']), label='Caffeine Intake', color='red')
plt.title('Productivity vs Caffeine Intake Over Time')
plt.xlabel('Date')
plt.ylabel('Normalized Scale (0–1)')
plt.xticks(rotation=45)
plt.legend()
plt.show()


![Productivity vs Caffeine Intake Over Time](figures/productivity_vs_caffeine_ot.jpg)




###Productivity vs Total Sleep Time Over Time
This line graph visualizes the relationship between daily productivity and total sleep time over the observation period. The blue line represents productivity, while the green line represents total sleep time. Both of the variables are normalized to 0-1 scale again to allow visual comparison despite their differing units. 

Looking at the plot, it can be observed that productivity and total sleep duration generally coincide with each other, though the alignment is less consistent than with caffeine intake, suggesting that while sleep duration may influence productivity, the effect appears moderate and not perfectly linear.
The plot also shows that the sleep duration varied considerably over the month, indicating that sleep patterns were not stable, probably highly due to external factors such as upcoming exams or assignments etc. 



In [ ]:
# Productivity vs Total Sleep Time Over Time
plt.figure(figsize=(10,5))
plt.plot(df['DATE'], normalize(df['Productivity_min']), label='Productivity')
plt.plot(df['DATE'], normalize(df['TotalSleep_min']), label='Total Sleep Time', color='yellowgreen')
plt.title('Productivity vs Total Sleep Time Over Time')
plt.xlabel('Date')
plt.ylabel('Normalized Scale (0–1)')
plt.xticks(rotation=45)
plt.legend()
plt.show()


![Productivity vs Total Sleep Time Over Time](figures/productivity_vs_totalsleep_ot.jpg)




## 5. Summary

This exploratory data analysis investigated the relationship between daily productivity, caffeine intake, and sleep patterns over a one-month observation period. By cleaning and organizing the dataset, we were able to compute summary statistics, visualize distributions, and examine correlations to gain insights into how these factors interact.

### Data Cleaning and Preprocessing: 
- Empty columns were removed and column names were standardized for clarity.
- Caffeine intake values were cleaned by removing “mg” strings and converting the values to numeric.
- Dates were parsed and sorted to allow chronological analysis.
### General Statistics: 
- The average daily caffeine intake was approximately 115 mg, with high variability (0–410 mg).
- Total sleep averaged 439 minutes (~7.3 hours), with moderate variability and a roughly symmetric distribution.
- REM, Light, and Deep sleep stages showed different levels of stability, with Deep sleep being the most consistent one.
- Productivity varied widely (60–360 minutes), showing considerable day-to-day fluctuations.
### Correlation Analysis: 
- Productivity was strongly positively correlated with caffeine intake (r = 0.83) and moderately correlated with total sleep (r = 0.55).
- Positive correlations also existed between productivity and REM and Light sleep although moderately, while Deep sleep showed a weak negative correlation.
- Based on the visualizations and the correlation matrix, there appears to be a pattern where higher caffeine consumption and longer sleep durations coincide with higher productivity, with caffeine showing a more pronounced trend. However, formal hypothesis testing is needed to confirm any significant relationships.
### Visualizations: 
- Histograms highlighted the right-skewed distribution of caffeine intake and relatively symmetric sleep duration. 
- Scatter plots suggested a positive trend between caffeine intake and productivity, and a moderate trend between sleep duration and productivity. 
- Line plots of normalized values over time illustrated that peaks in productivity often coincided with higher caffeine intake, while the relationship with total sleep was less consistent. 
